### PR TITLE
Tutorial Doc

### LINKED ISSUE #467: Document how to design a show from scratch
I'm working to understand how this all works, design a show, and contribute some docs improvements and have some questions.

To understand the basic concepts I'm trying to design a simple show from scratch, based on pieces from the "Pasture Bedtime" template.

For the show I'd like two things:

- "Solid Color" paint shader with two controls:
  - Color slider
  - Brightness slider
- "Scanner" filter shader

Screen Shot 2022-07-29 at 8 13 35 PM

> Stripped down "Pasture Bedtime" show shows a solid color with a vertical scan line!

## Some general questions...

What is the preferred terminology?

- "scenes" vs "paint filters"
- "controls" vs "gadgets"

What are the main UI components?

- Is far left model preview / Open / Show GLSL section always there? Why would I place a "visualizer" in the layout instead?
- What is a "vacuity"?
- What is the "UV In" xy control?

## Building a show

The basic steps are:

- `+ New Show` -> Empty Show
- Design Mode
- `+` -> New Button -> New Paint Shader

    ```cpp
    // Solid Color
    uniform vec4 color; // @@ColorPicker

    // @return color
    // @param uvIn uv-coordinate
    vec4 main(vec2 uvIn) {
        return vec4(color.rgb, 1.);
    }
    ```

  - Shader Name: Solid Color
  - Apply, then Close

- `+` -> New Button -> New Filter Shader

    ```cpp
    // Scanner
    uniform float time;

    // @param fragCoord uv-coordinate
    // @return color
    vec4 upstreamColor(vec2 fragCoord);

    // @return color
    // @param uvIn uv-coordinate
    // @param inColor color
    vec4 main(vec2 uvIn) {
        vec4 c = upstreamColor(uvIn);
        float width = .025;

        float xScan = cos(time) / 2. + .5;
        if (abs(uvIn.x - xScan) < width) {
            c.rgb *= .5;
            c.r = 0.; // 1. - c.r;
        }

        float yScan = sin(time) / 2. + .5;
        if (abs(uvIn.y - yScan) < width) {
            c.rgb *= .5;
            c.g = 0.; //  = 1. - c.g;
        }

        return vec4(c.r, c.g, c.b, 1.);
    }
    ```

- Save Show As... "tutorial.sparkle"

I have two main issues:

1) Placing controls is inconsistent in the UI
2) The shaders don't work as expected

## Placing Controls

The "Unplaced Controls" widget is inconsistent. Sometimes it doesn't show the unplaced controls. Sometimes the drag and drop from the window into the button grid drops the widget immediately in a fixed location. You can overlap buttons and have a hard time hitting the delete or edit.

But eventually I can get it to work with some combination of:

- exit design mode
- reload web browser
- tweak the name of the control variables
- drag a control
- move it to another area in the button grid
- drag another control

## Shaders

The "Solid Color" shader is working as expected! Next step is to extend it to have a alpha or brightness slider.

The "Scanner" shows a warning symbol and doesn't render a horizontal and vertical scan line on the model as it does in the "Pasture Bedtime" show. It does periodically change the brightness of the entire model.

Is there something more to do with ports? Or with the shader code?

Screen Shot 2022-07-29 at 8 36 27 PM

> Rebuilt show shows a solid color but no vertical or horizontal scan line...

### PR DESCRIPTION
**Pull request by nzoschke**
Document how to build a simple show from scratch. Fixes #467.

### CODE REVIEW

File: Tutorial.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+## Scene Quick Start
+
+Load a scene from the BAAAHS template. Defining a scene's model, light controllers and fixtures is an exercise for another day...
+
+### Load the BAAAHS template
+
+- "☰" to open the Show/Scene menu then Scene tab
+- "+ New Scene"
+- From template "BAAAHS" to see the sheep model 🐑
```
Comment (xian) [inline Tutorial.md:35]: I'd suggest using the "Hi-Res" template over BAAAHS. On the simulator, BAAAHS is pretty slow cuz the fake IP stack is (very) poorly optimized for lots of little packets. Hi Res is roughly the same number of pixels but way more responsive.

There's currently a sort-of bug in the template that makes Hi-Res look really dark which I'll fix. (I think the grid w/h just needs to be smaller; LEDs are rendered at a constant size in the visualizer and they're just too spread out right now.)

File: Tutorial.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+### Define a default projection
+
+- "☰" to open the Show/Scene menu then Show tab
+- Toggle "Design Mode"
+- "✎" next to the show name
+- New Projection Shader
+- Shader name: "Flat Projection"
+- Use the default code
+
+  ```c
```
Comment (xian) [inline Tutorial.md:58]: `glsl` probably works here too.

File: Tutorial.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+- "+ New Show"
+- "Empty Show"
+- "Save" button, enter "my-baaahs", then "save"
+
+### Define a default projection
+
+- "☰" to open the Show/Scene menu then Show tab
+- Toggle "Design Mode"
+- "✎" next to the show name
+- New Projection Shader
```
Comment (xian) [inline Tutorial.md:54]: Maybe we just put this one into the shader library?

File: Tutorial.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  // @return uv-coordinate
+  // @param pixelLocation xyz-coordinate
+  vec2 main(vec3 pixelLocation) {
+      vec3 start = modelInfo.center - modelInfo.extents / 2.;
+      vec3 rel = (pixelLocation - start) / modelInfo.extents;
+      return rel.xy;
+  }
+  ```
+- "Apply" then "Close"
```
Comment (xian) [inline Tutorial.md:73]: Ugh yeah sorry about that. Think I should add an "Apply and Close" button? It's so clunky.

File: Tutorial.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+- "☰" to open the Show/Scene menu then Show tab
+- "+ New Show"
+- "Empty Show"
+- "Save" button, enter "my-baaahs", then "save"
+
+### Define a default projection
+
+- "☰" to open the Show/Scene menu then Show tab
+- Toggle "Design Mode"
```
Comment (xian) [inline Tutorial.md:52]: Keyboard shortcut: `d`

File: Tutorial.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+### Create an empty show
+
+- "☰" to open the Show/Scene menu then Show tab
+- "+ New Show"
+- "Empty Show"
+- "Save" button, enter "my-baaahs", then "save"
+
+### Define a default projection
+
+- "☰" to open the Show/Scene menu then Show tab
```
Comment (xian) [inline Tutorial.md:51]: Keyboard shortcut: `Esc`

File: Tutorial.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      return vec4(color.rgb, 1.);
+  }
+  ```
+
+- "Apply" then "Close"
+- "Save"
+
+#### Controls
+
+The above effect needs a `ColorPicker`. While "Design Mode" has some support for seeing unplaced controls and placing them in the button grid, at present it is easiest to do this by editing the show template directly.
```
Comment (xian) [inline Tutorial.md:101]: Ouch. I'll try to improve on that.

The other possibility here is to create a Vacuity, the color picker should magically appear there.

**Review by xian — APPROVED** [review]
<3 <3 <3

File: Tutorial.md (hunk 2 above)
Comment (nzoschke) [reply Tutorial.md:58]: `glsl` this doesn't work in vscode out of the box but it does work on GitHub https://gist.github.com/nzoschke/937e91c0fd2fefb7e1fccb8ba5813a56

Do you have a preference either way?

Comment (xian) [reply Tutorial.md:58]: Not particularly, it'll do glsl keyword highlighting but no strong preference.

File: Tutorial.md (hunk 7 above)
Comment (nzoschke) [reply Tutorial.md:101]: Yeah this feels like the biggest wart as an end user right now. 

I'll try vacuity next.

And certainly no rush to fix it on my part, understanding and documenting workarounds is easy enough.

i'm sure you're plenty busy with other priorities :)

File: Tutorial.md (hunk 1 above)
Comment (nzoschke) [reply Tutorial.md]: Ok this definitely runs a lot faster! Thanks for the tip.

File: Tutorial.md (hunk 3 above)
Comment (nzoschke) [reply Tutorial.md:67]: Makes sense to me! Tried a quick spike on this but might have broken things. I'll follow up on moving this out of docs and into the library in a separate PR.

```diff
diff --git a/src/commonMain/resources/shader-libraries/built-in/_libraryIndex.json b/src/commonMain/resources/shader-libraries/built-in/_libraryIndex.json
index 85d9dbaaa..ecd8da297 100644
--- a/src/commonMain/resources/shader-libraries/built-in/_libraryIndex.json
+++ b/src/commonMain/resources/shader-libraries/built-in/_libraryIndex.json
@@ -3,6 +3,17 @@
   "description": "Some built-in shaders!",
   "license": null,
   "entries": [
+    {
+      "id": "flatprojection",
+      "title": "Flat Projection",
+      "description": "Project pixels onto a model",
+      "lastModifiedMs": 0,
+      "tags": [
+        "@filter",
+        "@type=projection"
+      ],
+      "srcFile": "projections/Flat.glsl"
+    },
     {
       "id": "beatshift",
       "title": "Beat Shift",
diff --git a/src/commonMain/resources/shader-libraries/built-in/projections/Flat.glsl b/src/commonMain/resources/shader-libraries/built-in/projections/Flat.glsl
new file mode 100644
index 000000000..c372856b2
--- /dev/null
+++ b/src/commonMain/resources/shader-libraries/built-in/projections/Flat.glsl
@@ -0,0 +1,17 @@
+// Flat Projection
+
+// Use this to project pixels onto a model
+
+struct ModelInfo {
+    vec3 center;
+    vec3 extents;
+};
+uniform ModelInfo modelInfo;
+
+// @return uv-coordinate
+// @param pixelLocation xyz-coordinate
+vec2 main(vec3 pixelLocation) {
+    vec3 start = modelInfo.center - modelInfo.extents / 2.;
+    vec3 rel = (pixelLocation - start) / modelInfo.extents;
+    return rel.xy;
+}
```

```
Missing incoming links on shader "Flat Projection
No link for ports: modelInfo, pixelLocation
```

Comment (nzoschke) [reply Tutorial.md:67]: oh maybe I didn't have the a scene loaded

Comment (nzoschke) [reply Tutorial.md]: @xian ok moved this to the library! LMK if I did it correctly... Could also split this out into its own PR if we want to keep this docs only.

File: Tutorial.md (hunk 2 above)
Comment (nzoschke) [reply Tutorial.md]: going to stich with `c` for now

File: Tutorial.md (hunk 7 above)
Comment (nzoschke) [reply Tutorial.md]: vacuity works and is much easier! will update docs to this

File: Tutorial.md (hunk 4 above)
Comment (nzoschke) [reply Tutorial.md:69]: It's not so bad... Given that you can make changes on other tabs in the popup a single "Save" that closes might be annoying too.

**Comment by nzoschke** [comment]
Final mystery is the scanner has a warning... Where can I see the warning and how to resolve... 🤔

Screen Shot 2022-08-06 at 12 01 54 PM

**Comment by xian** [comment]
It should show up in the list of all warnings for the show, which is a pop-up when you click on the warning icon next to "Sparkle Motion" in the header bar.

I bet the error is something to do with patch ports. Click the "Ports" tab on the shader editor and scroll down.

**Review by xian — APPROVED** [review]

**Comment by nzoschke** [comment]
Maybe "Unknown Port | Old Source" in the ports section? I'll debug this in follow up.

**Pull request merged**